orchestrator/graphql/utils/modify_class.py

```diff
@@ -0,0 +1,121 @@
+# Copyright 2019-2023 SURF.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import inspect
+import types
+from collections.abc import Callable, Iterable, Sequence
+from enum import IntEnum, StrEnum
+from functools import partial
+from typing import Any, get_args, get_origin
+
+import strawberry
+from more_itertools import consume, first_true, side_effect
+from pydantic import create_model
+from pydantic.fields import FieldInfo
+from strawberry.experimental.pydantic.conversion_types import StrawberryTypeFromPydantic
+
+from nwastdlib import const
+from orchestrator.domain.base import DomainModel
+from orchestrator.types import is_list_type
+from pydantic_forms.types import strEnum
+
+MatchFunc = Callable
+ActionFunc = Callable
+ModifyMap = Iterable[tuple[MatchFunc, ActionFunc]]
+
+
+def is_subclass_of(annotation: type, class_or_tuple: type | tuple[type]) -> bool:
+    return inspect.isclass(annotation) and issubclass(annotation, class_or_tuple)
+
+
+is_domain_model = partial(is_subclass_of, class_or_tuple=DomainModel)
+is_int_enum = partial(is_subclass_of, class_or_tuple=IntEnum)
+is_str_enum = partial(is_subclass_of, class_or_tuple=(strEnum, StrEnum))
+
+
+def is_optional_enum_of_type(annotation: type | None, is_type_enum: Callable) -> bool:
+    origin = get_origin(annotation)
+    if origin is not None and origin is types.UnionType:
+        args = get_args(annotation)
+        return len(args) == 2 and is_type_enum(args[0]) and args[1] is type(None)
+    return False
+
+
+is_optional_int_enum = partial(is_optional_enum_of_type, is_type_enum=is_int_enum)
+is_optional_str_enum = partial(is_optional_enum_of_type, is_type_enum=is_str_enum)
+
+
+def map_type(annotation: type | None, modify_map: ModifyMap) -> type | None:
+    match first_true(modify_map, pred=lambda match: match[0](annotation)):
```

Comment:
do you expect extra cases? if not it could just be an if statement

Reply:
The match is a nice way to do the destructuring of the return value (a tuple). Pattern often used in Scala and other languages :)